### PR TITLE
mesa: fix missing timespec_get on darwin

### PR DESCRIPTION
**Pull request by bobrik**
###### Motivation for this change

Similar to #113751.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS (aarch64)
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/mesa/default.nix [hunk 1]
```diff
@@ -65,6 +65,11 @@ stdenv.mkDerivation {
       url = "https://gitlab.freedesktop.org/mesa/mesa/commit/aebbf819df6d1e.patch";
       sha256 = "17248hyzg43d73c86p077m4lv1pkncaycr3l27hwv9k4ija9zl8q";
     })
+    # See: https://gitlab.freedesktop.org/mesa/mesa/-/issues/1020
+    (fetchpatch {
+      url = "https://raw.githubusercontent.com/XQuartz/XQuartz/master/src/mesa/mesa.patches/0004-Hack-to-address-build-failure-when-using-newer-macOS.patch";
```
Comment (primeos) [inline pkgs/development/libraries/mesa/default.nix:70]: This should be e.g.: https://raw.githubusercontent.com/XQuartz/XQuartz/91f6b3d98ece6d566aa9f52b85aca1785f84f88e/src/mesa/mesa.patches/0004-Hack-to-address-build-failure-when-using-newer-macOS.patch

Links to the `master` branch aren't stable and the content could change at any time.

**Review by primeos — CHANGES_REQUESTED** [review]
This PR needs to target `staging` (I can help you with that if you want).

And could you please document the reason for this PR a bit better (also in the commit message)? I.e. does this fix the build on `aarch64-darwin` or is it something else (/more).

And did you verify that this doesn't break `x86_64-darwin`?
If so it should be fine.

File: pkgs/development/libraries/mesa/default.nix (hunk 1 above)
Comment (bobrik) [reply pkgs/development/libraries/mesa/default.nix]: Updated to use a file on disk to avoid this issue.

**Comment by bobrik** [comment]
> This PR needs to target `staging` (I can help you with that if you want).

I've updated it to only target `aarch64-darwin` stdenv to avoid any rebuilds any allow it to target `master` instead. Let me know if you'd still like me to retarget `staging`.

> And could you please document the reason for this PR a bit better (also in the commit message)? I.e. does this fix the build on aarch64-darwin or is it something else (/more).

Expanded the comment and the a commit message.

> And did you verify that this doesn't break x86_64-darwin?

Not even touching `x86_64-darwin` with the latest update to be 100% sure.

File: pkgs/development/libraries/mesa/default.nix (hunk 1 above)
Comment (primeos) [reply pkgs/development/libraries/mesa/default.nix]: Ok. AFAIK we prefer `fetchpatch` (https://nixos.org/manual/nixpkgs/unstable/#sec-patches, but IIRC not everyone agrees with that) and my link above should be stable but in this case adding the patch to Nixpkgs should be fine as the patch is small and we'd have to fetch it from a different repository (which the maintainers could theoretically delete or force-push to - shouldn't be a problem in this case though).

**Review by primeos — APPROVED** [review]
> I've updated it to only target aarch64-darwin stdenv to avoid any rebuilds any allow it to target master instead.

Ok, LGTM then. Thanks :)

File: pkgs/development/libraries/mesa/default.nix [hunk 2]
```diff
@@ -65,6 +65,10 @@ stdenv.mkDerivation {
       url = "https://gitlab.freedesktop.org/mesa/mesa/commit/aebbf819df6d1e.patch";
       sha256 = "17248hyzg43d73c86p077m4lv1pkncaycr3l27hwv9k4ija9zl8q";
     })
+  ] ++ optionals (stdenv.isDarwin && stdenv.isAarch64) [
+    # Fix aarch64-darwin build, remove when upstreaam supports it out of the box.
+    # See: https://gitlab.freedesktop.org/mesa/mesa/-/issues/1020
+    ./aarch64-darwin.patch
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/mesa/default.nix:71]: Can this be applied on all platforms? If you please remove the conditional.

Comment (primeos) [reply pkgs/development/libraries/mesa/default.nix:71]: See https://github.com/NixOS/nixpkgs/pull/113754#issuecomment-782726994. If it works on `x86_64-darwin` I could also make it unconditional during the next regular update.

**Pull request merged**